examples/sparse/graphsage.py

```diff
@@ -0,0 +1,236 @@
+    model.eval()
+    with torch.no_grad():
+        logits = model(A, features)
+        logits = logits[mask]
+        labels = labels[mask]
+        _, indices = torch.max(logits, dim=1)
+        correct = torch.sum(indices == labels)
+        return correct.item() * 1.0 / len(labels)
+
+
+def train(A, features, labels, masks, model):
+    # Define train/val samples, loss function and optimizer.
+    train_mask, val_mask = masks
+    loss_fcn = nn.CrossEntropyLoss()
+    optimizer = torch.optim.Adam(model.parameters(), lr=1e-2, weight_decay=5e-4)
+
+    # Training loop.
+    for epoch in range(100):
+        model.train()
+        logits = model(A, features)
+        loss = loss_fcn(logits[train_mask], labels[train_mask])
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        acc = evaluate(A, features, labels, val_mask, model)
+        print(
+            f"Epoch {epoch:05d} | Loss {loss.item():.4f} | Accuracy {acc:.4f} "
+        )
+
+
+class SAGEConv(nn.Module):
+    r"""GraphSAGE layer from `Inductive Representation Learning on
+    Large Graphs <https://arxiv.org/pdf/1706.02216.pdf>`__
+    """
+
+    def __init__(
+        self,
+        in_feats,
+        out_feats,
+        aggregator_type="gcn",
+        feat_drop=0.0,
+        bias=True,
+        norm=None,
+        activation=None,
+    ):
+        super(SAGEConv, self).__init__()
+        self._in_src_feats, self._in_dst_feats = in_feats, in_feats
+        self._out_feats = out_feats
+        self._aggre_type = aggregator_type
+        self.norm = norm
+        self.feat_drop = nn.Dropout(feat_drop)
+        self.activation = activation
+
+        # aggregator type: mean/gcn
+        self.fc_neigh = nn.Linear(self._in_src_feats, out_feats, bias=False)
+
+        if aggregator_type != "gcn":
+            self.fc_self = nn.Linear(self._in_dst_feats, out_feats, bias=bias)
+        elif bias:
+            self.bias = nn.parameter.Parameter(torch.zeros(self._out_feats))
+        else:
+            self.register_buffer("bias", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        gain = nn.init.calculate_gain("relu")
+        if self._aggre_type != "gcn":
+            nn.init.xavier_uniform_(self.fc_self.weight, gain=gain)
+        nn.init.xavier_uniform_(self.fc_neigh.weight, gain=gain)
+
+    def forward(self, A, feat):
+        feat_src = feat_dst = self.feat_drop(feat)
+        feat_dst = feat_src[: A.shape[0]]
+
+        h_self = feat_dst
+        lin_before_mp = self._in_src_feats > self._out_feats
+        srcdata, dstdata = {}, {}
+        if self._aggre_type == "mean":
+            srcdata["h"] = (
+                self.fc_neigh(feat_src) if lin_before_mp else feat_src
+            )
+            # Div by degree
+            D_hat = dglsp.diag(A_hat.sum(0)) ** -1
+            A_div = A_hat @ D_hat
+            dstdata["neigh"] = A_div.T @ srcdata["h"]
+            h_neigh = graph.dstdata["neigh"]
```

Comment:
Where is `graph` from?

Reply:
It is a typo mistake, I have fixed this.